### ISSUE TITLE
Memory usage increases significantly per character in Entry

### ISSUE THREAD

**Issue by VideoCarp** [fyne-io/fyne#2969]
Apparently, this is not a bug. It is still important to point it out, though.
And I can’t delete the issue, anyway.

After searching, I could not find anything related to this.
### Describe the bug:

`Entry` seems to have some problems with memory. Adding any character drastically increased memory usage.  
  
The memory usage increases extremely quickly when characters are added to a `widget` `Entry`. I added the string "Hello" in
a text entry. The memory usage jumped from a stable 14.2 MB to a stable 18 MB. I left it for a while and the memory was still high.
I also gave it 100 characters. The memory went from 14.2 MB to a stable 22.5 MB. 
And giving it 245, it managed to reach 110 MB of memory usage. But this was a multiline entry.

### To Reproduce:
Steps to reproduce the behaviour:
Use this code: https://paste.gg/p/anonymous/b92b73bf8870438895df687f2a7984d1
which creates a `widget.NewEntry` in a `container.New` with a `layout.NewGridLayout`.
Run `go mod init`
Run `go mod tidy`
Compile with `go build -ldflags -H=windowsgui` or equivalent.

### Screenshots:

![image](https://user-images.githubusercontent.com/66365570/167392419-471fbfec-e88e-4e6c-a128-213dff86dec0.png)
![image](https://user-images.githubusercontent.com/66365570/167392650-86edb657-5f8f-44d1-99d4-4550d08aa86a.png)
![image](https://user-images.githubusercontent.com/66365570/167393575-86c93fc6-fb26-4bbe-a26a-5285be15a80b.png)
![image](https://user-images.githubusercontent.com/66365570/167401262-98aa0009-8f3b-484b-8c42-c2786adb5970.png)

The last image has a roughly 245 characters. Though, it is of a `MultiLineEntry`
(Note: I restarted the app for every picture). Sorry if the images aren't clear, I don't know why that suddenly happened. They should be readable, though.

### Example code:

```go
app := app.New()
win := app.NewWindow("")
content := container.New(
    widget.NewEntry(),
)
win.SetContent(content)
win.ShowAndRun()
```

### Device (please complete the following information):
 - **OS:**  Windows
 - **Version:**  10
 - **Go version:**  1.18.1
 - **Fyne version:**  v2.1.4

**Comment by andydotxyz** [comment]
This relates, in part, to other issues related to texture cache like #348, but with text it is worse because we cache based on complete text, which as you are typing will almost never hit the cache.
In 2.3 we are working towards a better text implementation which will slice into words for texture cache so this should be improved soon.

**Comment by luyaotang** [comment]
等待2.3

**Comment by andydotxyz** [comment]
This remains a high priority and has been enabled by changes to our text rendering, but will not be ready for 2.4.0 - moving to top of the list for 2.4.1

**Comment by steampoweredtaco** [comment]
I am writing a debug tool which will create large multiline entries and ran into performance issues that I think fall in this umbrella, but I characterize this issue with very poor performance and haven't recorded the memory usage.

I can file a new bug if you think I should.  This is on Windows 10 + 2.5.3-rc1 (and lower), go 1.23.3.

The scenario is a large amount of texts in an entry and only after setting `.Multiline = true` and `.Wrapping = fyne.TextWrapWord`

Minimal reproduction I have is:
```go
package main

import (
	"fmt"
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/widget"
	"log"
	"math/rand"
	"os"
	"runtime/pprof"
	"time"
)

const (
	// At about 2K ~2s, 4K ~10s, 8K ~37-40s
	low  = 8000
	high = 8001
)

// This adds breakable characters as well to show it isn't about
// how long it takes to wrap but in general how much text there is.
func randString(low, high int) string {
	cnt := low + rand.Intn(high-low)
	s := make([]byte, cnt)
	for i := range s {
		s[i] = byte(rand.Intn('}'-' ') + ' ')
	}
	return string(s)
}

func main() {
	f, err := os.Create("cpuprof")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()
	pprof.StartCPUProfile(f)
	defer pprof.StopCPUProfile()
	a := app.New()
	w := a.NewWindow("slow render")
	w.Resize(fyne.NewSize(400, 400))
	e := widget.NewEntry()
	e.Text = randString(low, high)
	// If you comment out the next two lines there isn't an issue with
	// performance showing it has something to do with the multiline/wrap
	e.MultiLine = true
	e.Wrapping = fyne.TextWrapWord
	scroll := container.NewVScroll(e)
	// Adding the scroll content is enough to show the performance issue with large wrapped text
	n := time.Now()
	w.SetContent(scroll)
	fmt.Print("time to set content:", time.Since(n))
	// sometimes it takes the same amount of time to resize, but not always, left this out
	// w.Resize(fyne.NewSize(400, 400))
	go func() {
		time.Sleep(2 * time.Second)
		// Show no issues with resizing the window after initial load
		for i := 400.; i <= 800.; i += .5 {
			// not nearly as bad as setContent but still very slow and can start getting flickery at high counts
			w.Resize(fyne.NewSize(float32(i), 400))
		}
		w.Close()
	}()
	w.ShowAndRun()
}
```
See comments for more details of the issue.  At 8K characters it can take upwards of 40 seconds to set the content of a window with this entry inside of a scroll.  It gets worst from there and you can easily change the above program to experiment.

This is the cpu profile at 8K characters.  [cpuprof.zip](https://github.com/user-attachments/files/18002028/cpuprof.zip)

A non trivial amount of time is spent in runtime slicerunetostring, but the majority seems to be throughout glDriver RenderTextSize stack and the harfbuzz algorithm.

This video (first ~37 seconds waiting for it to load) shows also the flickering/slowness of the program quickly trying to resize the width of the window containing a large entry.
  
https://github.com/user-attachments/assets/ac3f7c78-a004-406e-8eef-f38dbc517d19

**Comment by andydotxyz** [comment]
Thanks @steampoweredtaco but this ticket is about memory usage and your report seems to be about CPU utilisation so I think they are different. Memory is about cache usually but CPU probably means our wrapping algorithms need work...

**Comment by steampoweredtaco** [comment]
Fair. Prior to the latest RC, there was more time spent in cache management
and why it seemed related initially. However, it seems to have shifted
after rc1.

I'll create a new issue later today, thanks.

**Comment by andydotxyz** [comment]
I tested this on latest develop and, although upon initial entry memory can go up a little I was able to paste 100s of characters over many lines and it only went up by another 0.2MB
